orderbook.go

```diff
@@ -0,0 +1,181 @@
+package main
+
+const MAX_PRICE = 10000000
+
+type PricePoint struct {
+    orderHead *Order
+    orderTail *Order
+}
+
+func (pp *PricePoint) Insert(o *Order) {
+    if pp.orderHead == nil {
+        pp.orderHead = o
+        pp.orderTail = o
+    } else {
+        pp.orderTail.next = o
+        pp.orderTail = o
+    }
+}
+
+func (pp *PricePoint) Peek() *Order {
+    return pp.orderHead
+}
+
+func (pp *PricePoint) Pop() *Order {
+    order := pp.orderHead
+    pp.orderHead = order.next
+    // Pop the last node
+    if order.next == nil {
+        pp.orderTail = nil
+    }
+    return order
+}
+
+type OrderStatus int
+const (
+    OS_NEW OrderStatus = iota
+    OS_OPEN
+    OS_PARTIAL
+    OS_FILLED
+    OS_CANCELLED
+)
+
+type Order struct {
+    id uint64
+    isBuy bool
+    price uint32
+    amount uint32
+    filled uint32
+    status OrderStatus
+    next *Order
+}
+
+func NewOrder(id uint64, isBuy bool, price uint32, amount uint32) *Order {
+    o := new(Order)
+    o.id = id
+    o.isBuy = isBuy
+    o.price = price
+    o.amount = amount
+    o.status = OS_NEW
+    return o
+}
+
+type OrderBook struct {
+    // These are more estimates than reportable values
+    ask uint32
+    bid uint32
+    orderIndex map[uint64]*Order
+    prices [MAX_PRICE]*PricePoint
```

Comment:
Well this doesn't scale very well. Maybe they should be buckets, or something. But the traditional approach is a pair of ordered linked lists, buys and sells.